plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/PassthroughQueryPageSource.java

```diff
@@ -41,40 +53,55 @@ public PassthroughQueryPageSource(ElasticsearchClient client, ElasticsearchTable
         long start = System.nanoTime();
         result = client.executeQuery(table.getIndex(), table.getQuery().get());
         readTimeNanos = System.nanoTime() - start;
+
+        if (result.v2()) {
+            this.iterator = new SearchHitIterator(client, result.v1());
+        }
     }
 
     @Override
     public long getCompletedBytes()
     {
-        return result.length();
+        return totalBytes;
     }
 
     @Override
     public long getReadTimeNanos()
     {
-        return readTimeNanos;
+        return readTimeNanos + (iterator != null ? iterator.getReadTimeNanos() : 0);
     }
 
     @Override
     public boolean isFinished()
     {
-        return done;
+        return (iterator != null && !iterator.hasNext()) || done;
     }
 
     @Override
     public Page getNextPage()
     {
-        if (done) {
-            return null;
-        }
+        if (!done) {
```

Comment:
it looks like we handle two codepaths in these methods one for scrolling and non-scrolling case. (in getReadTimeNanos, isFinished and getNextPage)

IMO the flow would be simpler using an interface which'll have scrolling and non-scrolling implementations. (for doScroll=true and false respectively)

i.e. 

```
    private static abstract class SearchHitIterator
            extends AbstractIterator<String> implements Closeable
    {
        public abstract long getReadTimeNanos();
    }
```